File: WikiCodeParser/Elements/Element.cs
using System;
using System.Collections.Generic;
using System.Linq;
using WikiCodeParser.Nodes;

namespace WikiCodeParser.Elements
{
    public abstract class Element
    {
        public List<string> Scopes { get; set; }
        public int Priority { get; set; } = 0;

        public virtual bool InScope(string scope) => string.IsNullOrWhiteSpace(scope) ||
                                                     Scopes.Contains(scope, StringComparer.InvariantCultureIgnoreCase);

        public abstract bool Matches(Lines lines);
        public abstract INode Consume(Parser parser, Lines lines);
    }
}

File: php/src/Elements/Element.php
<?php

namespace LogicAndTrick\WikiCodeParser\Elements;

use LogicAndTrick\WikiCodeParser\Lines;
use LogicAndTrick\WikiCodeParser\Nodes\INode;
use LogicAndTrick\WikiCodeParser\ParseData;
use LogicAndTrick\WikiCodeParser\Parser;

abstract class Element
{
    /**
     * @var string[]
     */
    public array $scopes = [];

    public int $priority = 0;

    public function InScope(string $scope) : bool {
        return !$scope || trim($scope) == '' || in_array($scope, $this->scopes, true);
    }

    public abstract function Matches(Lines $lines) : bool;
    public abstract function Consume(Parser $parser, ParseData $data, Lines $lines, string $scope) : ?INode;
}